### PR TITLE
MongoDB Change Streams (async generators approach), common fixes and improvements

### PR DESCRIPTION
**Pull request by vladislav-shkutovich**

### CODE REVIEW

File: src/user/user.service.ts [hunk 1]
```diff
@@ -126,36 +118,88 @@ export class UserService implements OnModuleInit {
...
+        sort: { updatedAt: -1 },
+        limit: USER_RECENT_TODOS_COUNT,
+      })
+  
+      return await this.userDatabaseService.updateUser(userId, {
+        todos: recentUserTodos,
+      })
+      */
+    }
   }
```
Comment (vladislav-shkutovich) [inline src/user/user.service.ts:156]: Just highlighting of this comment:
```// ? Question: access DB for deprecated user.todos by `getUserById` to update them manually, or access DB for user.todos directly by the `getUserTodos` using query params? In both approaches we access DB, so why not the second one? In both options we access for the same data.```

We already discussed it before, but as I remebmer, the main goal of returning the `reduce` method is an accessing the DB less times. But in both of these options we access the DB at least 1 time. So, the Option 2 looks quite sufficient here, isn't it?

File: src/todos/todos.database.service.ts [hunk 2]
```diff
@@ -18,7 +18,9 @@ export class TodosDatabaseService {
     userId: string,
   ): Promise<Todo> {
     const createdTodo = await this.todoModel.create({
-      userId,
+      // ? Question: It's a temporal fix, but need to discuss a problem with so much conversions from string into ObjectId and back again. I left some paper notes about it, let's discuss it on the next call.
+      // ? This topic is also related to the IdParamDto discussion.
+      userId: new Types.ObjectId(userId),
       ...createTodoDto,
     })
     return createdTodo.toObject()
```
Comment (vladislav-shkutovich) [inline src/todos/todos.database.service.ts]: Just highlighting of this comment to discuss it on the next call

File: src/user/user.service.ts [hunk 3]
```diff
@@ -126,36 +125,86 @@ export class UserService implements OnModuleInit {
...
+        sort: { updatedAt: -1 },
+        limit: USER_RECENT_TODOS_COUNT,
+      })
+  
+      return await this.userDatabaseService.updateUser(userId, {
+        todos: recentUserTodos,
+      })
+      */
+    }
+  }
```
Comment (vladislav-shkutovich) [inline src/user/user.service.ts]: Just highlighting of this comment to discuss it on the next call

// ? Question: access DB for deprecated user.todos by `getUserById` to update them manually, or access DB for user.todos directly by the `getUserTodos` using query params? In both approaches we access DB, so why not the second one? In both options we access for the same data.

We already discussed it before, but as I remebmer, the main goal of returning the reduce method is an accessing the DB less times. But in both of these options we access the DB at least 1 time. So, the Option 2 looks quite sufficient here, isn't it?

File: src/common/dto/id-param.dto.ts [hunk 4]
```diff
@@ -3,6 +3,10 @@ import { IsMongoId } from 'class-validator'
 export class IdParamDto {
   @IsMongoId()
   // TODO: - Transform string into ObjectId in commonly used IdParamDto and update related code (lot of conversions);
-  // @Transform()
+  // ? To be discussed.
+  // @Transform(({ value }) =>
+  //   typeof value === 'string' ? new Types.ObjectId(value) : value,
+  // )
+  // id: Types.ObjectId
```
Comment (vladislav-shkutovich) [inline src/common/dto/id-param.dto.ts]: Just highlighting of this comment to discuss it on the next call

File: types/mongodb/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+import type { ChangeStream as MongodbChangeStream } from 'mongodb'
+
+declare module 'mongodb' {
+  interface ChangeStream extends MongodbChangeStream {
+    driverChangeStream: ChangeStream<TSchema, TChange<TSchema>>
+  }
+}
```
Comment (vladislav-shkutovich) [inline types/mongodb/index.d.ts]: These types looks a bit weird and actually it's almost useless, but I just skipped this after some time spent on it. Let's discuss it on the next call

File: src/todos/todos.change-stream.database.service.ts [hunk 6]
```diff
@@ -57,17 +63,15 @@ export class TodosChangeStreamDatabaseService {
...
-        const updatedTodo = changeStreamDoc.fullDocument as Todo
-        callback(updatedTodo)
-      },
-    )
+    for await (const changeStreamDoc of this.changeStreamOnUpdate
+      .driverChangeStream) {
+      yield changeStreamDoc.fullDocument
+    }
+
+    this.changeStreamOnUpdate.close()
```
Comment (masontikhonov) [inline src/todos/todos.change-stream.database.service.ts]:
```suggestion
    await this.changeStreamOnUpdate.close()
```

File: src/todos/todos.change-stream.database.service.ts [hunk 7]
```diff
@@ -78,27 +82,11 @@ export class TodosChangeStreamDatabaseService {
...
+      .driverChangeStream) {
+      yield changeStreamDoc.documentKey._id
+    }
 
-        if (userWithDeletedTodo) {
-          callback(userWithDeletedTodo.todos[0])
-        }
-      },
-    )
+    this.changeStreamOnDelete.close()
```
Comment (masontikhonov) [inline src/todos/todos.change-stream.database.service.ts]:
```suggestion
    await this.changeStreamOnDelete.close()
```

File: src/todos/todos.change-stream.database.service.ts [hunk 8]
```diff
@@ -34,16 +41,15 @@ export class TodosChangeStreamDatabaseService {
...
-        const createdTodo = changeStreamDoc.fullDocument as Todo
-        callback(createdTodo)
-      },
-    )
+    for await (const changeStreamDoc of this.changeStreamOnCreate
+      .driverChangeStream) {
+      yield changeStreamDoc.fullDocument
+    }
+
+    this.changeStreamOnCreate.close()
```
Comment (masontikhonov) [inline src/todos/todos.change-stream.database.service.ts]:
```suggestion
    await this.changeStreamOnCreate.close()
```

File: src/todos/todos.change-stream.database.service.ts [hunk 9]
```diff
@@ -8,22 +8,29 @@ import type {
...
+  >
 
   constructor(
     @InjectModel(TODO_MODEL) private readonly todoModel: Model<Todo>,
-    @InjectModel(USER_MODEL) private readonly userModel: Model<User>,
   ) {}
 
-  subscribeOnTodoCreate(callback: (todo: Todo) => void): void {
+  async *subscribeOnTodoCreate() {
     if (!this.changeStreamOnCreate) {
```
Comment (masontikhonov) [inline src/todos/todos.change-stream.database.service.ts:34]:
```suggestion
  async *subscribeOnTodoCreate(): AsyncGenerator<Todo> {
```

File: src/todos/todos.change-stream.database.service.ts [hunk 10]
```diff
@@ -34,16 +41,15 @@ export class TodosChangeStreamDatabaseService {
...
+      .driverChangeStream) {
+      yield changeStreamDoc.fullDocument
+    }
+
+    this.changeStreamOnCreate.close()
   }
 
-  subscribeOnTodoUpdate(callback: (todo: Todo) => void): void {
+  async *subscribeOnTodoUpdate() {
     if (!this.changeStreamOnUpdate) {
```
Comment (masontikhonov) [inline src/todos/todos.change-stream.database.service.ts:53]:
```suggestion
  async *subscribeOnTodoUpdate(): AsyncGenerator<Todo> {
```

File: src/todos/todos.change-stream.database.service.ts [hunk 11]
```diff
@@ -57,17 +63,15 @@ export class TodosChangeStreamDatabaseService {
...
+      yield changeStreamDoc.fullDocument
+    }
+
+    this.changeStreamOnUpdate.close()
   }
 
-  // TODO: - Get only todoId in subscribeOnTodoDelete and send it into new UserService method (without changes as it was before);
-  subscribeOnTodoDelete(callback: (todo: Todo) => void): void {
+  async *subscribeOnTodoDelete() {
     if (!this.changeStreamOnDelete) {
```
Comment (masontikhonov) [inline src/todos/todos.change-stream.database.service.ts:75]:
```suggestion
  async *subscribeOnTodoDelete(): AsyncGenerator<ObjectId> {
```

**Pull request merged**